assistant: add file contents retrieval and basic text search tools

sharon-wang opened the pull request:
## Summary

- addresses https://github.com/posit-dev/positron/issues/7110

- adds tools to:
    - search for specific content and return a set of matches (does a search in the workspace for the text or regex provided)
    - get the contents for a specific file (file needs to be in the workspace or open in an editor)

### Release Notes

#### New Features

- Positron Assistant now has tools to get the workspace project tree, search for text in the workspace, and get the contents of a specific file (#7110)

### QA Notes

@:assistant

- Some test cases: empty workspace, multi-root workspace, single-root workspace
- Text search can be exact text match, word match, regex, etc.

sharon-wang committed:
add basic text search tool

sharon-wang committed:
add new `source` property to the text search tool data

sharon-wang committed:
enable edit file tool for positron assistant

sharon-wang committed:
add file contents assistant tool

sharon-wang committed:
Revert "enable edit file tool for positron assistant"

This reverts commit 249c35837feea170336739d6a6260785d0a4f048.

Seems like we need to do a bit more to get this tool to work.

Getting `edits not iterable` error.

seeM reviewed (commented):
Code changes look good! In my simple test, the model invoked the file contents tool with an invalid path. Can we improve that?

### QA feedback

I tried this out in the qa-example-content repo with the `pydotSample.py` file open.

I asked "tell me more about my project" and the model invoked the project tree tool

I then asked "what is the example.dot file?" referring to a file that the model just referenced. It invoked the file contents tool but with the wrong path:

<img width="346" alt="image" src="https://github.com/user-attachments/assets/b4df72dc-8bbf-449c-bcd8-cfc8d3783d49" />

I then asked "which files in my project use the pydot package?" and the model invoked the text search tool followed by the file contents tool, which seems reasonable!

---

BTW, we may want to use `prepareInvocation` and return `pastTenseMessage` to display a nicer message during invocation. E.g. see the `getPlot` tool in the `positron-assistant` extension. Not sure if there is any other way to change that.

It looks like `toolResultMessage` is displayed in the UI when the tool completes, and if it's not set the default is `Used "Tool"...` which IMO isn't great. We may want to refine our usage of `toolResultMessage`. I see the project tree tool doesn't specify it.

The text search tool used the default text in my test. I assume that means `messages` was empty. Maybe that's not the best place to put `messages`? Maybe it should be in the tool result content?

sharon-wang committed:
register tools iteratively so less boilerplate is needed

sharon-wang committed:
don't return toolResultMessage

sharon-wang committed:
set invocation message and past tense message for workspace tools

sharon-wang committed:
add text search result messages to tool output content

sharon-wang committed:
mark unused params in project tree tool invoke method

sharon-wang committed:
expand capabilities of text search tool

IPatternInfo properties can now be set, to allow for regular expression search, word matching, case sensitivity, etc.

sharon-wang committed:
improve system prompt and tool descriptions for better file path handling

sharon-wang commented on src/vs/workbench/contrib/chat/browser/tools/textSearchTool.ts:
```diff
@@ -0,0 +1,98 @@
+	source: { type: 'internal' },
+	modelDescription: findTextInProjectModelDescription,
+	tags: ['positron-assistant'],
+	canBeReferencedInPrompt: false,
```
Ah, that's what that means! I'll open a separate issue to add these to the "Add context" options, as it looks like we'l want an icon for each tool as well

sharon-wang commented on src/vs/workbench/contrib/chat/browser/tools/textSearchTool.ts:
```diff
@@ -0,0 +1,98 @@
+		};
+
+		const content = {
+			pattern: textToFind
```
I've modified this tool to support regex, case sensitivity and word-matching, along with a few other optional things.

Globs are not supported, but I think that will be more applicable to "Find files by glob patterns" in https://github.com/posit-dev/positron/issues/7297

sharon-wang reviewed (commented):
thanks for the great feedback @seeM!

here are some changes I've made
- added the tool invocation and past tense messages to the workspace tools
- added to the chat participant system prompt for better handling of file paths
- added more detail to the tool descriptions, also to help with file path handling and influence when the tools are called
- expanded the inputs to the text search tool, so that the llm can leverage regex search, specify case sensitivity, word matching, etc.
- removed the tool result message and instead pass the text search result messages as part of the tool output content
- small enhancement to the tools contribution registration code, so it is less boilerplatey to add a new tool

I used similar steps to your QA notes and I think the experience is a bit better now. Let me know what you think!

seeM commented on extensions/positron-assistant/src/participants.ts:
```diff
@@ -424,6 +424,10 @@ abstract class PositronAssistantParticipant implements IPositronAssistantPartici
 class PositronAssistantChatParticipant extends PositronAssistantParticipant implements IPositronAssistantParticipant {
 	id = ParticipantID.Chat;
+
+	protected override async getSystemPrompt(request: vscode.ChatRequest): Promise<string | undefined> {
```
I wonder if we also want this in the `EditParticipant` (used in the chat pane in edit mode)?

seeM commented on extensions/positron-assistant/src/participants.ts:
```diff
@@ -424,6 +424,10 @@ abstract class PositronAssistantParticipant implements IPositronAssistantPartici
 	id = ParticipantID.Chat;
+
+	protected override async getSystemPrompt(request: vscode.ChatRequest): Promise<string | undefined> {
+		return await fs.promises.readFile(`${mdDir}/prompts/chat/filepath.md`, 'utf8');
```
```suggestion
		return await fs.promises.readFile(`${mdDir}/prompts/chat/filepaths.md`, 'utf8');
```

seeM reviewed (dismissed):
Nice, my previous QA example works well now, and the tool usage texts look better.

Another small improvement, which could be done separately, is to add the referenced files to the chat response like the edit file tool does. It might be too heavy handed for the project tree tool since every file is referenced, but would definitely make sense for the file contents tool and perhaps also text search.

seeM commented:
Oh, I approved but you'll need to at least fix the `filepath.md` typo before merging.

sharon-wang committed:
use filepaths prompt for ask and edit chat participants

austin3dickey commented:
Ah, I just saw that on my PR too.
- https://github.com/posit-dev/positron/pull/7337#issuecomment-2830704142
